Shared/Headers/FocusEventListener.hpp

```diff
@@ -11,71 +11,89 @@
     std::function<void(TPayload &payload)> _onMessage;
     std::unique_ptr<std::thread> _eventListenerThread;
-    std::shared_ptr<zmq::socket_t> _socketSUB = std::make_shared<zmq::socket_t>(*FocusSocket::Context, ZMQ_SUB);
+    std::unique_ptr<zmq::socket_t> _socketSUB = std::make_unique<zmq::socket_t>(*FocusSocket::Context, ZMQ_SUB);
+    std::atomic<bool> _isRunning;
 
-    static void RunReceive(zmq::socket_t *socketSUB, const std::function<void(TPayload &)> onMessage) {
-        while (true) {
+    void RunReceive(zmq::socket_t *socketSUB, const std::function<void(TPayload &)> onMessage) {
+        while (_isRunning) {
             zmq::message_t msg;
-            socketSUB->recv(&msg);
-            socketSUB->recv(&msg);
-            std::string payload = std::string(static_cast<char*>(msg.data()), msg.size());
-            Focus::Event event;
-            if (!event.ParseFromString(payload))
-                continue;
-            onMessage(event);
+            if (socketSUB->recv(&msg)) {
+                if (!socketSUB->recv(&msg))
+                    continue;
+                std::string payload = std::string(static_cast<char *>(msg.data()), msg.size());
+                Focus::Event event;
+                if (!event.ParseFromString(payload))
+                    continue;
+                onMessage(event);
+            }
         }
     }
 
-    static void RunReceiveEnvelope(zmq::socket_t *socketSUB, const std::function<void(TPayload &)> onMessage) {
-        while (true) {
+    void RunReceiveEnvelope(zmq::socket_t *socketSUB, const std::function<void(TPayload &)> onMessage) {
+        while (_isRunning) {
             zmq::message_t msg;
-            socketSUB->recv(&msg);
-            socketSUB->recv(&msg);
-            std::string payload = std::string(static_cast<char*>(msg.data()), msg.size());
-            Focus::Envelope envelope;
-            if (!envelope.ParseFromString(payload))
-                continue;
-            onMessage(envelope);
+            if (socketSUB->recv(&msg)) {
+                if (!socketSUB->recv(&msg))
+                    continue;
+                std::string payload = std::string(static_cast<char *>(msg.data()), msg.size());
+                Focus::Envelope envelope;
+                if (!envelope.ParseFromString(payload))
+                    continue;
+                onMessage(envelope);
+            }
         }
     }
 
-    static void RunReceiveMessage(zmq::socket_t *socketSUB, const std::function<void(TPayload &)> onMessage) {
-        while (true) {
+    void RunReceiveMessage(zmq::socket_t *socketSUB, const std::function<void(TPayload &)> onMessage) {
+        while (_isRunning) {
             zmq::message_t msg;
-            socketSUB->recv(&msg);
-            socketSUB->recv(&msg);
-            std::string payload = std::string(static_cast<char*>(msg.data()), msg.size());
-            onMessage(payload);
+            if (socketSUB->recv(&msg)) {
```

Comment:
You can replace `zmq::message_t` by a `zmq::multipart_t`. The `recv` will block until all messages are received.